openhcl/virt_mshv_vtl/src/processor/tdx/mod.rs

```diff
@@ -373,6 +375,105 @@ impl VirtualRegister {
     }
 }
 
+/// Interface for managing lower VTL timer deadlines via TDX L2-VM TSC Deadline
+/// Timer capability.
+///
+/// This allows VTL2 to set an execution deadline for lower VTLs, in absolute
+/// virtual TSC units. If the lower VTL is running when the deadline time
+/// arrives, it exits to VTL2 with exit reason `VmxExitBasic::TIMER_EXPIRED`.
+/// If the TSC deadline is in the past during entry into lower VTL (i.e., TSC
+/// deadline value is lower than the current virtual TSC value), it will immediately
+/// exit back to VTL2 with exit reason `VmxExitBasic::TIMER_EXPIRED`.
+///
+/// The TSC deadline is set using `TDG.VP.WR` for `TDVPS.TSC_DEADLINE[L2-VM Index]`.
+/// The actual `TDG.VP.WR` call to set the deadline is made by the `mshv_vtl` driver
+///  before entering the lower VTL.
+struct TdxTscDeadlineService {
+    // Fixed-point scale factor to convert 100ns to TSC units.
+    tsc_scale_100ns: u128,
+}
+
+impl TdxTscDeadlineService {
+    /// Convert hypervisor reference time (in 100ns) to TSC units.
+    fn ref_time_to_tsc(&self, ref_time: u64) -> u64 {
+        // Use fixed-point multiplication to calculate:
+        // tsc_ticks = (time_100ns /  10_000_000) * tsc_frequency
+        ((ref_time as u128 * self.tsc_scale_100ns) >> 64) as u64
+    }
+
+    /// Returns true if `ref_time` is before `ref_time_last`.
+    ///
+    /// Note that this is a relative comparison in the 64-bit space and is not
```

Comment:
I'm not sure i understand the implications of this comment? or how this check cannot be transitive?